include/crow/routing.h

```diff
@@ -950,83 +998,60 @@ namespace crow
-            {
-                if (n->param_childrens[i])
-                {
-                    CROW_LOG_DEBUG << std::string(2*level, ' ') /*<< "("<<n->param_childrens[i]<<") "*/;
-                    switch(static_cast<ParamType>(i))
+                    if (!piece_found)
                     {
-                        case ParamType::INT:
-                            CROW_LOG_DEBUG << "<int>";
-                            break;
-                        case ParamType::UINT:
-                            CROW_LOG_DEBUG << "<uint>";
-                            break;
-                        case ParamType::DOUBLE:
-                            CROW_LOG_DEBUG << "<float>";
-                            break;
-                        case ParamType::STRING:
-                            CROW_LOG_DEBUG << "<str>";
-                            break;
-                        case ParamType::PATH:
-                            CROW_LOG_DEBUG << "<path>";
-                            break;
-                        default:
-                            CROW_LOG_DEBUG << "<ERROR>";
-                            break;
+                        auto new_node_idx = new_node(idx);
+                        new_node_idx->key = c;
+                        idx = new_node_idx;
                     }
-
-                    debug_node_print(&nodes_[n->param_childrens[i]], level+1);
                 }
             }
-            for(auto& kv : n->children)
-            {
-                CROW_LOG_DEBUG << std::string(2*level, ' ') /*<< "(" << kv.second << ") "*/ << kv.first;
-                debug_node_print(&nodes_[kv.second], level+1);
-            }
-        }
 
-    public:
-        void debug_print()
-        {
-            debug_node_print(head(), 0);
+            //check if the last node already has a value (exact url already in Trie)
+            if (idx->rule_index)
+                throw std::runtime_error("handler already exists for " + url);
+            idx->rule_index = rule_index;
         }
 
-    private:
-        const Node* head() const
+        size_t get_size()
         {
-            return &nodes_.front();
+            return get_size(&head_);
         }
 
-        Node* head()
+        size_t get_size(Node* node)
         {
-            return &nodes_.front();
+            unsigned size = 8; //rule_index and param
+            size += (node->key.size()); //each character in the key is 1 byte
+            for (auto child: node->children)
+            {
+                size += get_size(child);
+            }
+            return size;
         }
 
-        unsigned new_node()
+
+    private:
+
+        Node* new_node(Node* parent)
         {
-            nodes_.resize(nodes_.size()+1);
-            return nodes_.size() - 1;
+            auto& children = parent->children;
+            children.resize(children.size()+1);
+            children[children.size()-1] = new Node();
```

Comment:
Maybe smart pointer?

Reply:
ngl I'm not sure how smart pointers work, I'll look into and implement them

Reply:
I looked into implementing it, it seems to be inflexible (I couldn't set it to `head` if `find()`). And I don't believe normal pointers should be a problem since the only possible change happens on startup and the only possible place where a pointer would be inaccessible is also where it's deleted.

Reply:
I meant e.g. std::shared_ptr ;) But raw pointer shouldn't be a problem here.

Reply:
Understood, I'll merge this for now and look into `shared_ptr`